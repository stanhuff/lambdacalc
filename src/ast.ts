import { ParseError } from './parserUtils';
import { panic } from "tse-utils/assertUtils";

export abstract class Node {
}

export enum EvaluationType {
    substitute,
    reduction,
    none
}

export class Evaluation {
    constructor(
        readonly type: EvaluationType,
        readonly result: Expression,
    ) { }
}

export abstract class Expression extends Node {
    abstract substitute(variables: Scope): Expression;
    reduce(): Expression { return this; }
    abstract format(): string;
    abstract evaluateOnce(variables: Scope): Evaluation;
    abstract evaluate(variables: Scope): Expression;
}

export class Variable extends Expression {
    constructor(
        readonly name: string,
    ) { super(); }

    substitute(variables: Scope): Expression {
        const newNode = variables[this.name];
        return newNode ? newNode : this;
    }

    format() {
        return this.name;
    }

    evaluateOnce(variables: Scope): Evaluation {
        return new Evaluation(EvaluationType.none, this);
    }

    evaluate(variables: Scope): Expression {
        return this;
    }
}

export class Lambda extends Expression {
    constructor(
        readonly variable: string,
        readonly body: Expression,
    ) { super(); }

    substitute(variables: Scope): Expression {
        const newBody = this.body.substitute(this.excludeBoundVariables(variables));
        return (newBody == this.body) ? this : new Lambda(this.variable, newBody);
    }

    apply(value: Expression) {
        const variables = { [this.variable]: value };
        return this.body.substitute(variables);
    }

    private excludeBoundVariables(variables: Scope): Scope {
        if (this.variable in variables) {
            const newVariables = Object.assign({}, variables);
            delete newVariables[this.variable];
            return newVariables;
        }
        return variables;
    }

    evaluateOnce(variables: Scope): Evaluation {
        /* simplify lambda
        if (!(this.body instanceof Variable)) {
            const evaluation = this.body.evaluateOnce(variables);
            if (evaluation.result !== this.body)
                return new Evaluation(evaluation.type, new Lambda(this.variable, evaluation.result));
        }
        */
        return new Evaluation(EvaluationType.none, this);
    }

    evaluate(variables: Scope) {
        return this;
    }

    format() {
        return `Î»${this.variable}. ${this.body.format()}`;
    }
}

export interface Scope {
    [index: string]: Expression;
}

export class Call extends Expression {
    constructor(
        readonly func: Expression,
        readonly arg: Expression,
    ) { super(); }

    evaluateOnce(variables: Scope): Evaluation {
        if (this.arg instanceof Call) {
            const evaluation = this.arg.evaluateOnce(variables);
            return new Evaluation(evaluation.type, new Call(this.func, evaluation.result));
        }

        if (this.func instanceof Call) {
            const evaluation = this.func.evaluateOnce(variables);
            return new Evaluation(evaluation.type, new Call(evaluation.result, this.arg));
        }
        else if (this.func instanceof Lambda) {
            return new Evaluation(EvaluationType.reduction, this.func.apply(this.arg));
        }
        else {
            const newFunc = this.func.substitute(variables);
            return (newFunc === this.func)
                ? new Evaluation(EvaluationType.none, this)
                : new Evaluation(EvaluationType.substitute, new Call(newFunc, this.arg));
        }
    }

    evaluate(variables: Scope): Expression {
        const arg = this.evaluateArg(variables);

        const func = this.evaluateFunc(variables);

        if (func instanceof Lambda)
            return func.apply(arg);
        else
            return func === this.func && arg === this.arg ? this : new Call(func, arg);
    }

    private evaluateArg(variables: Scope) {
        let arg = this.arg;
        while (arg instanceof Call)
            arg = arg.evaluate(variables);
        return arg;
    }

    private evaluateFunc(variables: Scope) {
        let func = this.func;
        while (!(func instanceof Lambda)) {
            const next = func instanceof Call ? func.evaluate(variables) : func.substitute(variables);
            if (next === func)
                break;
            func = next;
        }
        return func;
    }

    substitute(variables: Scope): Expression {
        const newFunc = this.func.substitute(variables);
        const newArg = this.arg.substitute(variables);
        return (newFunc === this.func && newArg === this.arg) ? this : new Call(newFunc, newArg);
    }

    reduce(): Expression {
        const func = this.func.reduce();
        if (func instanceof Lambda)
            return func.apply(this.arg.reduce());
        return this;
    }

    format(hasRight?: boolean) {
        const left = this.formatChild(this.func, true);
        const right = this.formatChild(this.arg, hasRight || false);
        return this.arg instanceof Call ? `${left} (${right})` : `${left} ${right}`;
    }

    private formatChild(expr: Expression, hasRight: boolean): string {
        if (expr instanceof Call)
            return expr.format(hasRight);
        else
            return hasRight && expr instanceof Lambda ? `(${expr.format()})` : expr.format();
    }
}

export interface IO {
    output(text: string): void;
    showSteps?: boolean;
}

export abstract class Statement extends Node {
    abstract execute(scope: Scope, io: IO): void;
}

export class StatementList extends Statement {
    statements: Statement[];
    constructor(...statements: Statement[]) {
        super();
        this.statements = statements.slice();
    }

    execute(scope: Scope, io: IO) {
        this.statements.forEach(v => v.execute(scope, io));
    }
}

export class Assignment extends Statement {
    constructor(
        readonly variable: string,
        readonly initializer: Expression
    ) { super(); }

    execute(scope: Scope, io: IO) {
        scope[this.variable] = this.initializer;
    }
}

export class ExpressionStatement extends Statement {
    constructor(
        readonly node: Expression
    ) { super(); }

    execute(scope: Scope, io: IO): void {
        if (io.showSteps) {
            let last = this.node;
            io.output(this.getStepType(EvaluationType.none) + last.format());
            while (true) {
                const evaluation = last.evaluateOnce(scope);
                if (evaluation.type === EvaluationType.none)
                    break;
                last = evaluation.result;
                io.output(this.getStepType(evaluation.type) + last.format());
            }
        }
        else {
            let last = this.node;
            while (true) {
                const next = last.evaluate(scope);
                if (next === last)
                    break;
                last = next;
            }
            io.output(last.format());
        }
    }

    private getStepType(evaluationType: EvaluationType): string {
        switch (evaluationType) {
            case EvaluationType.none:
                return "    ";
            case EvaluationType.reduction:
                return "--> ";
            case EvaluationType.substitute:
                return " =  ";
            /* istanbul ignore next */
            default:
                return panic(`Unexpected EvaluationType: ${EvaluationType[evaluationType]}`)
        }
    }
}

export class AstError extends Error {
}