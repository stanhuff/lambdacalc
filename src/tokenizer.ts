import { ParseError } from "./parserUtils";

const SPACE = 0x20,
    TAB = 0x9,
    NL = 0xA,
    CR = 0xD,
    UPPER_A = 'A'.charCodeAt(0),
    LOWER_A = 'a'.charCodeAt(0),
    UPPER_Z = 'Z'.charCodeAt(0),
    LOWER_Z = 'z'.charCodeAt(0),
    ZERO = '0'.charCodeAt(0),
    NINE = '9'.charCodeAt(0);

export enum TokenType {
    PERIOD = 46,
    LAMBDA = 0x3bb,
    EOF = -1,
    NUM = -2,
    LPAREN = 40,
    RPAREN = 41,
    WHITESPACE = -3,
    ID = -4,
    EQUALS = '='.charCodeAt(0),
    SEMICOLON = ';'.charCodeAt(0)
}

export class Token {
    constructor(
        public readonly type: TokenType,
        public readonly source: string,
        public readonly start: number,
        public readonly stop: number
    ) { }

    get text() { return this.source.substring(this.start, this.stop); }
}

export function getTokenTypeName(tokenType: TokenType) {
    if (tokenType >= 0)
        return `'${String.fromCodePoint(tokenType)}'`;
    else
        return TokenType[tokenType];
}

abstract class TokenClass {
    abstract tokenType: TokenType;
    abstract isCodePoint(codePoint: number, idx: number): boolean;
}

class IdTokenClass extends TokenClass {
    tokenType = TokenType.ID;
    isCodePoint(codePoint: number) {
        return codePoint >= UPPER_A && codePoint <= UPPER_Z || codePoint >= LOWER_A && codePoint <= LOWER_Z;
    }
}

class NumberTokenClass extends TokenClass {
    tokenType = TokenType.NUM;
    isCodePoint(codePoint: number) {
        return codePoint >= ZERO && codePoint <= NINE;
    }
}

class WhitespaceTokenClass {
    tokenType = TokenType.WHITESPACE;
    isCodePoint(codePoint: number) {
        return codePoint === SPACE || codePoint === TAB || codePoint === NL || codePoint === CR;
    }
}

export class Tokenizer {
    private idx: number = 0;

    constructor(
        private text: string
    ) { }

    private static tokenClasses: TokenClass[] = [new IdTokenClass(), new NumberTokenClass(), new WhitespaceTokenClass()];

    *getTokens() {
        for (; ;) {
            let token = this.getToken();
            if (token === undefined)
                break;
            yield token;
        }
        yield new Token(TokenType.EOF, this.text, this.idx, this.idx);
    }

    private getToken() {
        const codePoint = this.text.codePointAt(this.idx);
        return codePoint === undefined ? undefined : this.lexToken(codePoint);
    }

    private lexToken(codePoint: number) {
        switch (codePoint) {
            case TokenType.PERIOD:
            case TokenType.LAMBDA:
            case TokenType.LPAREN:
            case TokenType.RPAREN:
            case TokenType.EQUALS:
            case TokenType.SEMICOLON:
                const start = this.idx;
                const stop = ++this.idx;
                return new Token(codePoint, this.text, start, stop);
            default:
                const tokenClass = this.getTokenClass(codePoint);
                if (tokenClass)
                    return this.lexTokenClass(tokenClass);
                throw new ParseError(`Unexpected ${String.fromCodePoint(codePoint)}`);
        }
    }

    private getTokenClass(codePoint: number) {
        for (let i = 0; i < Tokenizer.tokenClasses.length; ++i) {
            if (Tokenizer.tokenClasses[i].isCodePoint(codePoint, 0)) {
                return Tokenizer.tokenClasses[i];
            }
        }
        return undefined;
    }

    private lexTokenClass(tokenClass: TokenClass) {
        const start = this.idx;
        let cp = this.text.codePointAt(this.idx);
        while (cp !== undefined && tokenClass.isCodePoint(cp, this.idx - start)) {
            this.idx += (isSurrogatePair(cp) ? 2 : 1);
            cp = this.text.codePointAt(this.idx);
        }
        const stop = this.idx;
        return new Token(tokenClass.tokenType, this.text, start, stop);
    }
}

function isSurrogatePair(codepoint: number) {
    return codepoint > 65535;
}