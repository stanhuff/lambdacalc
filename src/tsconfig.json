{
  "extends": "../base-tsconfig.json",
  "compilerOptions": {
    "outDir": "../build",
    "noEmitOnError": true,
    "module": "commonjs",    
    "types": [
      "node"
    ]
  },
  "exclude": [
    "**/__mocks__/**/*"
  ]
}
