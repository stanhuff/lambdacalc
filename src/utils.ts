
export function panic(message: string): never {
    throw new Error("Panic: " + message);
}

export function* filter<T>(itr: Iterable<T>, predicate: (v: T) => boolean) {
    for (const v of itr) {
        if (predicate(v))
            yield v;
    }
}

export function toArray<T>(itr: Iterable<T>): T[] {
    const ary: T[] = [];
    for (const n of itr)
        ary.push(n);
    return ary;
}

