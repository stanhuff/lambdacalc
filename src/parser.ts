import { Token, TokenType, getTokenTypeName } from "./tokenizer";
import { toArray, filter } from "./utils";
import { Node, Variable, Call, Lambda, StatementList, Assignment, ExpressionStatement, Statement, Expression } from './ast';
import { ParseError } from './parserUtils';

export class Parser {
    private tokens: Token[];
    private idx = -1;

    constructor(
        tokens: Iterable<Token>
    ) {
        this.tokens = toArray(filter(tokens, v => v.type !== TokenType.WHITESPACE));
    }

    parse(): Node | undefined {
        this.next();
        switch (this.token.type) {
            case TokenType.EOF:
                return undefined;
            default:
                return this.parseApply();
        }
    }

    parseStatementList(): StatementList {
        this.next();
        const statements = [];
        while (this.token.type !== TokenType.EOF) {
            statements.push(this.parseStatement());
        }
        return new StatementList(...statements);
    }

    private parseStatement(): Statement {
        const statement = (this.token.type == TokenType.ID && this.LA.type === TokenType.EQUALS)
            ? this.parseAssignmentStatement()
            : this.parseFunctionApplicationStatement();
        this.ensure(TokenType.SEMICOLON);
        return statement;
    }

    private parseAssignmentStatement() {
        const id = this.ensure(TokenType.ID);
        this.ensure(TokenType.EQUALS);
        const value = this.parseApply();
        return new Assignment(id.text, value);
    }

    private parseFunctionApplicationStatement() {
        const application = this.parseApply();
        return new ExpressionStatement(application);
    }

    private parseApply(): Expression {
        let func = this.parsePrimary();
        while (!this.isApplyTerminator()) {
            const arg = this.parsePrimary();
            func = new Call(func, arg);
        }
        return func;
    }

    private isApplyTerminator() {
        return this.token.type === TokenType.RPAREN || this.token.type === TokenType.SEMICOLON || this.token.type === TokenType.EOF
    }

    private parsePrimary(): Expression {
        switch (this.token.type) {
            case TokenType.LPAREN:
                this.consume();
                const term = this.parseApply();
                this.ensure(TokenType.RPAREN);
                return term;
            case TokenType.ID:
                return new Variable(this.consume().text);
            case TokenType.LAMBDA:
                return this.parseFunc();
            default:
                return this.reportError(`Unexpected token type ${getTokenTypeName(this.token.type)}.  Expected ${getTokenTypeName(TokenType.LPAREN)}, ${getTokenTypeName(TokenType.LAMBDA)}, or ${getTokenTypeName(TokenType.ID)}`, this.token);
        }
    }

    private parseFunc(): Expression {
        this.ensure(TokenType.LAMBDA);
        const paramToken = this.ensure(TokenType.ID);
        this.ensure(TokenType.PERIOD);
        const body = this.parseApply();
        return new Lambda(paramToken.text, body);
    }

    private get token() { return this.tokens[this.idx]; }
    private get LA() { return this.tokens[this.idx + 1]; }

    private next() {
        this.consume();
        return this.token;
    }

    private consume() {
        const token = this.token;
        do {
            this.idx++;
        } while (this.tokens[this.idx].type === TokenType.WHITESPACE);
        return token;
    }

    private ensure(tokenType: TokenType) {
        if (this.token.type === tokenType)
            return this.consume();
        return this.reportError(`Unexpected token type ${getTokenTypeName(this.token.type)}.  Expected ${getTokenTypeName(tokenType)}.`, this.token);
    }

    private reportError(message: string, token: Token): never {
        throw new ParseError(`Error: ${message} @ ${token.start}`);
    }
}
