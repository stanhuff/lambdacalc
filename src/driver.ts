import { IO } from './ast';
import { Tokenizer } from './tokenizer';
import { Parser } from './parser';

export function runScript(script: string, io: IO) {
    const t = new Tokenizer(script);

    const p = new Parser(t.getTokens());

    const sl = p.parseStatementList();

    const scope = {};

    sl.execute(scope, io);

}