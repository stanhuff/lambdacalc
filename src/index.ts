import * as fs from "fs";
import { runScript } from './driver';

runScript(
    fs.readFileSync(process.argv[2], "utf-8"),
    {
        output(v: string) {
            process.stdout.write(v);
            process.stdout.write("\n");
        }
    }
);
