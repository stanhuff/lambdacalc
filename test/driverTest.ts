import { runScript } from "../src/driver";

describe("driver", () => {

    test("simple", () => {

        let output = "";
        const io = { output: (v: string) => output += v };

        runScript("tru = λt. λf. t;\n        fls = λt. λf. f;\n        test = λl. λm. λn. l m n;\n\n       tru pass fail;\n        fls fail pass;\n\n        λy.y x;\n        tru;\n        test;\n\n        test tru v w;\n", io);

        expect(output).toMatchSnapshot();

    });

});