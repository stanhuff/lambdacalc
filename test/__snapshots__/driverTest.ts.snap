// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`driver simple 1`] = `"passpassλy. y xλt. λf. tλl. λm. λn. l m n(λt. λf. t) v w"`;
